### PR TITLE
Fix gdal_calc for use in modeler

### PR DESCRIPTION
**Pull request by rudivs**
Do not process optional parameters that are empty strings. Fixes #19329
(https://issues.qgis.org/issues/19329)

### CODE REVIEW

File: python/plugins/processing/algs/gdal/gdalcalc.py [hunk 1]
```diff
@@ -222,7 +222,7 @@ def getConsoleCommands(self, parameters, context, feedback, executing=True):
         if self.parameterAsString(parameters, self.BAND_A, context):
             arguments.append('--A_band ' + self.parameterAsString(parameters, self.BAND_A, context))
 
-        if self.INPUT_B in parameters and parameters[self.INPUT_B] is not None:
+        if self.INPUT_B in parameters and parameters[self.INPUT_B] is not (None or ''):
```
[the remark this answers is not in the document]
Comment (rudivs) [reply python/plugins/processing/algs/gdal/gdalcalc.py]: My bad. Would `if self.INPUT_B in parameters and not parameters[self.INPUT_B]:` be an acceptable test?

[the remark this answers is not in the document]
Comment (nyalldawson) [reply python/plugins/processing/algs/gdal/gdalcalc.py]: Only for 3.4. in 3.2 valid vector layers sometimes return false for the bool operator.

[the remark this answers is not in the document]
Comment (rudivs) [reply python/plugins/processing/algs/gdal/gdalcalc.py]: Could do `if self.INPUT_B in parameters and parameters[self.INPUT_B] is not None and parameters[self.INPUT_B] != ''` if there's no more concise option?

[the remark this answers is not in the document]
Comment (nyalldawson) [reply python/plugins/processing/algs/gdal/gdalcalc.py]: I think that's the way to go.

**Comment by nyalldawson** [comment]
We also need a test added for this to https://github.com/qgis/QGIS/blob/master/python/plugins/processing/tests/GdalAlgorithmsTest.py

**Comment by rudivs** [comment]
I tested with QGIS 2.18 and this is a regression. I think this issue might affect other algorithms in modeler view. I suspect the problem comes in here:

https://github.com/qgis/QGIS/blob/2695e000422b48b3772a39c7a4040f29d43d0d61/python/plugins/processing/gui/wrappers.py#L867-L870

Rather than updating gdalcalc, I propose rather to fix this section in `wrappers.py` with:

            if self.param.flags() & QgsProcessingParameterDefinition.FlagOptional:
                self.combo.addItem(self.NOT_SELECTED, self.NOT_SET_OPTION)
            for layer in layers:
                self.combo.addItem(self.dialog.resolveValueDescription(layer), layer)

I have tested this with gdalcalc and it also fixes #19329. It also seems to be more in keeping with how other combo boxes resolve the not-selected option (and `NOT_SELECTED = '[Not selected]'` is already added to the class but not used). However, I see that there has been some history on this so I thought I'd check first in case this would introduce some other kind of regression.

**Comment by luipir** [comment]
@rudivs I remember this problem was fixed (I fixed it) when I reintroduced gdal_calc... but I can be my fault... now focusing on other processing fix. can you check the PR (and discussion) where I reintroduced this algorithm?
Sorry I can't help more, trying to find a generic solution for native mapcalculator

**Comment by rudivs** [comment]
Thanks @luipir, I don't mind doing a little investigation. The PR when you reintroduced the algorithm was (https://github.com/qgis/QGIS/pull/6984). The test for `None` was introduced in https://github.com/qgis/QGIS/pull/6984/commits/39154f4411e1db8cc17fea1c38343cd225637907 (based on a suggested snippet from Nyall). There was no further discussion about handling optional rasters in that pull request. You later fixed the management of optional rasters in batch mode (https://github.com/qgis/QGIS/pull/7180). The solution there was to add a value test to `BatchInputSelectionPanel` (also a generalised solution for a problem found in gdalcalc). This wouldn't be a fix for using gdalcalc in a model.

I couldn't find anything else in the history that would affect optional layers in models. Does this match your memory, or do you think I might be missing some history?

**Comment by luipir** [comment]
tnaks for the investigatio... may "memory scope" is limited due the fact that I'm also processing data and not focusing only on code. BTW, I can't understand the running context. Are you saying that you fix works running the alg (only) inside a model?

**Comment by rudivs** [comment]
The issue that this fixes is only encountered when running this algorithm inside a model. There is no problem with running it standalone or as a batch process, so this fix would not affect those situations.

Also, I'm proposing that we don't go ahead with this fix, since it would be unnecessary if https://github.com/qgis/QGIS/pull/7362 is merged (which I think is a more general solution).

**Pull request closed without merging**